File: app/src/main/java/com/innoveworkshop/gametest/engine/GameSurface.java
package com.innoveworkshop.gametest.engine;

import android.content.Context;
import android.content.res.Resources;
import android.graphics.BitmapFactory;
import android.graphics.Canvas;
import android.graphics.Color;
import android.util.AttributeSet;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import androidx.annotation.NonNull;

import com.innoveworkshop.gametest.R;

import java.util.ArrayList;
import java.util.Timer;
import java.util.TimerTask;

public class GameSurface extends SurfaceView {
    private SurfaceHolder holder;
    private Timer timer;

    private GameObject root;
    private ArrayList<GameObject> gameObjects;

    public GameSurface(Context context) {
        this(context, null);
    }

    public GameSurface(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public GameSurface(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);

        // Create the GameObject list.
        gameObjects = new ArrayList<GameObject>();

        // Ensure we are on top of everything.
        setZOrderOnTop(true);

        // Set up the SurfaceHolder event handler.
        holder = getHolder();
        holder.addCallback(new SurfaceHolder.Callback() {
            @Override
            public void surfaceCreated(@NonNull SurfaceHolder holder) {
                // Ensure we get the onDraw events.
                setWillNotDraw(false);

                // Start up the root object.
                root.onStart(GameSurface.this);

                // Set up the fixed update timer.
                timer = new Timer();
                timer.scheduleAtFixedRate(new FixedUpdateTimer(), 0, 1000 / 30);
            }

            @Override
            public void surfaceChanged(@NonNull SurfaceHolder holder, int format, int width, int height) {
            }

            @Override
            public void surfaceDestroyed(@NonNull SurfaceHolder holder) {
                // TODO: Stop everything.
            }
        });
    }

    public void setRootGameObject(GameObject root) {
        this.root = root;
    }

    public void addGameObject(GameObject gameObject) {
        gameObjects.add(gameObject);
        gameObject.onStart(this);
    }

    public boolean removeGameObject(GameObject gameObject) {
        return gameObjects.remove(gameObject);
    }

    @Override
    protected void onDraw(@NonNull Canvas canvas) {
        super.onDraw(canvas);

        root.onDraw(canvas);
        for (GameObject gameObject : gameObjects) {
            gameObject.onDraw(canvas);
        }
    }

    class FixedUpdateTimer extends TimerTask {
        @Override
        public void run() {
            for (GameObject gameObject : gameObjects) {
                gameObject.onFixedUpdate();
            }

            root.onFixedUpdate();
            invalidate();
        }
    }
}

File: app/src/main/java/com/innoveworkshop/gametest/engine/GameSurface.kt
package com.innoveworkshop.gametest.engine

import android.content.Context
import android.graphics.Canvas
import android.util.AttributeSet
import android.view.SurfaceHolder
import android.view.SurfaceView
import java.util.Timer
import java.util.TimerTask

class GameSurface @JvmOverloads constructor(
    context: Context?,
    attrs: AttributeSet? = null,
    defStyleAttr: Int = 0
) : SurfaceView(context, attrs, defStyleAttr) {
    private val holder: SurfaceHolder
    private var timer: Timer? = null
    private var root: GameObject? = null

    // Create the GameObject list.
    private val gameObjects = ArrayList<GameObject>()

    init {
        // Ensure we are on top of everything.
        setZOrderOnTop(true)

        // Set up the SurfaceHolder event handler.
        holder = getHolder()
        holder.addCallback(object : SurfaceHolder.Callback {
            override fun surfaceCreated(holder: SurfaceHolder) {
                // Ensure we get the onDraw events.
                setWillNotDraw(false)

                // Start up the root object.
                root!!.onStart(this@GameSurface)

                // Set up the fixed update timer.
                timer = Timer()
                timer!!.scheduleAtFixedRate(FixedUpdateTimer(), 0, (1000 / 30).toLong())
            }

            override fun surfaceChanged(
                holder: SurfaceHolder,
                format: Int,
                width: Int,
                height: Int
            ) {
            }

            override fun surfaceDestroyed(holder: SurfaceHolder) {
                // TODO: Stop everything.
            }
        })
    }

    fun setRootGameObject(root: GameObject?) {
        this.root = root
    }

    fun addGameObject(gameObject: GameObject) {
        gameObjects.add(gameObject)
        gameObject.onStart(this)
    }

    fun removeGameObject(gameObject: GameObject): Boolean {
        return gameObjects.remove(gameObject)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        root!!.onDraw(canvas)
        for (gameObject in gameObjects) {
            gameObject.onDraw(canvas)
        }
    }

    internal inner class FixedUpdateTimer : TimerTask() {
        override fun run() {
            for (gameObject in gameObjects) {
                gameObject.onFixedUpdate()
            }

            root!!.onFixedUpdate()
            invalidate()
        }
    }
}
